docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,187 @@
+2. Hook execution can't be atomically linked to the pod's status: there will be a window where the pod is in a ready state, and a failed hook execution must revert the pod's state back to failed when it would be preferable to have never allowed to the pod to become ready prior to hook success.
+
+## Use cases
+
+1. As a MySQL user, I want my application deployment to seed the database with data the first time the database is used.
+2. As a MySQL user, I want slave instance deployments to be automatically registered with the master.
+3. As a MongoDB user, I want database peer instances to be automatically clustered during deployments.
+4. As a Rails application author, I want to perform a Rails database migration following an application deployment.
+
+#### Use-case: Seeding a MySQL database
+
+Databases are often initialized with example or other data the first time they are deployed. Users should be able to express a hook with causes the database to be initialized automatically.
+
+Consider an application deployment which uses a MySQL database via a service. In this use case, it's the responsibility of the application deployment to initialize the database.
+
+###### Container lifecycle hooks
+
+A container lifecycle hook associated with the application deployment suffers from race conditions: the hook would execute once per pod in the deployment, burdening the hook author to ensure the hook is idempotent and safely concurrent when faced with races to initialize the database.
+
+###### Post-deployment hooks
+
+A post-deployment hook could satisfy the use case by ensuring the initialization is performed only once per logical deployment.
+
+#### Use-case: Clustering a MySQL database
+
+MySQL requires an orchestrated workflow for slave cluster participation. For a slave to join the cluster, the master must be locked and exported, and slaves associated with the master. Users should be able to deploy clustered MySQL instances.
+
+###### Container lifecycle hooks
+
+Container lifecycle hooks can be applied to this use case. Post-start hooks attached to slave containers could attempt to acquire a lock on the master database (via services), and could then be initialized. The master locking provides a natural coordination mechanism for concurrent slave hook executions across the replica set.
+
+###### Post-deployment hooks
+
+Post-deployment hooks could also be used to satisfy the use case. A post deployment hook associated with the slave deployment configuration could acquire a single lock on the master database and initialize all slave instances that resulted from the deployment.
+
+#### Use-case: Clustering a MongoDB database
+
+MongoDB can be clustered with homogeneous peers which perform leader elections. Users should be able to define a MongoDB deployment in which all replicas are automatically clustered.
```

Comment:
Why is this not role assignment or nominal services?  Also, this has the same problem as the first one - it's an initialization time thing, not on rolling updates.